### PR TITLE
[foreman] increase timeout of foreman-debug to 15minutes

### PR DESCRIPTION
**Pull request by pmoravec**
foreman-debug (collecting lots of log files and mainly katello task
export) often runs longer than the default 5 minutes timeout.

Closes: #888.

Signed-off-by: Pavel Moravec <pmoravec@redhat.com>

### CODE REVIEW

**Comment by bmr-cymru** [comment]
Do we want any timeout at all for this command?

I've heard reports of `foreman-debug` running for in excess of 15 minutes (don't know if that's with current versions, or the older, buggy builds).

**Comment by pmoravec** [comment]
I have discussed the options (15 mins timeout or no timeout, that is the question) with few relevant devels and we made into conclusion some timeout is better choice. Despite the stuck `foreman-debug` shall not reoccur, still it makes sense to prevent it. The tool calls various commands that theoretically might hang (esp. when calling `sosreport` when having `foreman` in troubles). 15minutes timeout shall cover >99% of use cases - esp. due to the fact the longest command - task export - should collect a limited amount of tasks only since some recent changes.

So TL;DR version: the 15mins timeout is the preferred option.

**Comment by bmr-cymru** [comment]
Perhaps `foreman-debug` should be implementing its own timeout handling, in order to have better control?

**Comment by bmr-cymru** [comment]
> should collect a limited amount of tasks only since some recent changes.

Not something we can rely on much in `sos` - there are many users who grudgingly update components only when asked (unless we start requiring this fixed `foreman-debug` in future `sos` builds.. one option, I guess).

**Comment by pmoravec** [comment]
> Perhaps foreman-debug should be implementing its own timeout handling, in order to have better control?

@lzap (as `foreman-debug` maintainer), could you comment if that is possible/planned/.. ? I see it definitely technically possible, but quite nontrivial change (i.e. how to interrupt `foreman-rake` command collecting task export?).

**Comment by lzap** [comment]
Foreman-debug is a dumb shell script and it does not aim to replace sos, but to complement it (and allow some debug collection for our Debian/Ubuntu users). I don't think we should be implementing anything like that there. Sorry for the delay.

**Comment by bmr-cymru** [comment]
> and allow some debug collection for our Debian/Ubuntu users

The sos package has been included in Debian and Ubuntu for literally _years_ now, so that argument doesn't really fly!

**Comment by lzap** [comment]
I haven't noticed. Nice, but it is what it is. We started a tiny shell script and it's bigger now. One day we will hopefully provide native sos plugin. Thanks!

**Comment by lzap** [comment]
Isn't better to remove this plugin from core distribution and simply provide sosreport-foreman package with it? This will solve the issue we have when sosreport is released in a different pace than Foreman/Satellite 6.

**Comment by bmr-cymru** [comment]
> Isn't better to remove this plugin from core distribution and simply provide sosreport-foreman package with 
> it? This will solve the issue we have when sosreport is released in a different pace than Foreman/Satellite 6.

No. The RHEV team did this with the vdsm plugin - it's a disaster. We don't currently make any API compatibility guarantees which means that everything has to be updated in lockstep anyway, and the coordination becomes painful for everyone involved.

If/when we have completed the move to declarative (metadata only) plugins then that's something that we can reconsider, but while plugin files can contain complex Python code and interact with the whole `Plugin`/`sos` API it is not something we are prepared to support.

**Comment by lzap** [comment]
Hmm then I think having a separate tools is fine. Having this in the upstream sos distribution will not work at all, the project debug tool must be up to date with the project itself. We have a different pace. Also we need to have it on old distributions and our tool also runs on Solaris and FreeBSD which we support too.

Anyway, back to the business - if I follow your conversation correctly, you do not like increasing the value.

@pmoravec I think we should pull the long job that causes this (I think its job export?) out of the default run and only execute it with special flag. Increasing will just give us some time for breathing but not enough...

**Comment by pmoravec** [comment]
Re idea to let foreman-debug to skip calling the only lengthy command task export (from it's plugin "katello-debug"): on one side efficient solution, on the other side support representatives would have to ask for sosreport *and* task export separately.

What about to:
- skip calling task export from foreman-debug when being called from sosreport (i.e. when `-g` option is passed to foreman-debug, like sos does)
- add to foreman plugin new command to call task export with increased timeout (see https://github.com/Katello/katello-packaging/blob/master/katello/katello-debug.sh#L84-L91 esp. line 87)

Technically this requires changes in sos, foreman-debug (to pass `-g` to katello-debug) and in katello-debug, each change is trivial.

Does this seem reasonable compromise?

**Comment by bmr-cymru** [comment]
> Anyway, back to the business - if I follow your conversation correctly, you do not like increasing the value.

I'm not sure what gave that impression, but it is not the case - it is ugly, and a hack, but if it is the best solution available at the moment then it's what we need to do.

The whole reason we have timeouts is to improve the users's experience in cases like this - that is not helped if the timeout fires too early for something that is known to take a long time.

I do still think that this would work fine as a plugin in the upstream distribution if there was a bit more coordination - then the debug shell script really would be just a "simple, small" script that grabs any additional bits that were missed out from the last upstream version.

This was actually the basis on which we originally agreed to call these `xyz-debug` scripts (not `foreman` in the first case, but a similar situation) - what seems to have happened however is that the projects that started these have continued to add features there, and never ask for, or submit anything, for the `sos` plugin. Over time the scripts get more and more complex and featureful until they have their own plugin mechanisms etc. and are requiring increased timeouts to get their work done.

I *do* think that is an unfortunate development, and I *do* think it would be simpler for all concerned to minimize this, but I just look after `sos`: it's not my place to dictate that to other projects (I can just try to explain the benefits of doing so).

**Comment by lzap** [comment]
I agree with @bmr-cymru in this case, either sosreport or foreman-debug should give reasonable result in reasonable time. The task export thing, that has been added by katello plugin, is slow and I think it's the time to pull it off foreman-debug, or make it much much faster (e.g. fetching the relevant data via psql rather than booting whole Rails stack).

**Comment by bmr-cymru** [comment]
👍

**Comment by bmr-cymru** [comment]
OK for 3.4 we will go with the increased timeout - if a better solution becomes available down the line we can switch to it then.
